MediaBrowser.Common/Plugins/BasePluginOfT.cs

```diff
@@ -0,0 +1,231 @@
+            }
+        }
+
+        /// <summary>
+        /// Gets the application paths.
+        /// </summary>
+        /// <value>The application paths.</value>
+        protected IApplicationPaths ApplicationPaths { get; private set; }
+
+        /// <summary>
+        /// Gets the XML serializer.
+        /// </summary>
+        /// <value>The XML serializer.</value>
+        protected IXmlSerializer XmlSerializer { get; private set; }
+
+        /// <summary>
+        /// Gets the type of configuration this plugin uses.
+        /// </summary>
+        /// <value>The type of the configuration.</value>
+        public Type ConfigurationType => typeof(TConfigurationType);
+
+        /// <summary>
+        /// Gets or sets the event handler that is triggered when this configuration changes.
+        /// </summary>
+        public EventHandler<BasePluginConfiguration> ConfigurationChanged { get; set; }
+
+        /// <summary>
+        /// Gets the name the assembly file.
+        /// </summary>
+        /// <value>The name of the assembly file.</value>
+        protected string AssemblyFileName => Path.GetFileName(AssemblyFilePath);
+
+        /// <summary>
+        /// Gets or sets the plugin configuration.
+        /// </summary>
+        /// <value>The configuration.</value>
+        public TConfigurationType Configuration
+        {
+            get
+            {
+                // Lazy load
+                if (_configuration == null)
+                {
+                    lock (_configurationSyncLock)
+                    {
+                        if (_configuration == null)
+                        {
+                            _configuration = LoadConfiguration();
+                        }
+                    }
+                }
+
+                return _configuration;
+            }
+
+            protected set => _configuration = value;
+        }
+
+        /// <summary>
+        /// Gets the name of the configuration file. Subclasses should override.
+        /// </summary>
+        /// <value>The name of the configuration file.</value>
+        public virtual string ConfigurationFileName => Path.ChangeExtension(AssemblyFileName, ".xml");
+
+        /// <summary>
+        /// Gets the full path to the configuration file.
+        /// </summary>
+        /// <value>The configuration file path.</value>
+        public string ConfigurationFilePath
+        {
+            get
+            {
+                var dataFolderPath = Path.Combine(ApplicationPaths.PluginsPath, Path.GetFileNameWithoutExtension(AssemblyFilePath));
+                if (!Directory.Exists(dataFolderPath) && Version != null)
+                {
+                    // Try again with the version number appended to the folder name.
+                    return dataFolderPath + "_" + Version.ToString();
+                }
+
+                return dataFolderPath;
+            }
+        }
```

Comment:
```suggestion
        public string ConfigurationFilePath => Path.Combine(ApplicationPaths.PluginConfigurationsPath, ConfigurationFileName);
```

I think this was supposed to be reverted?

Reply:
That's the bug fix due to dataFolderPath now possibly being invalid due to our versioning change.

eg. dataFolderPath points to _/server/plugins/myplugin_

whereas we now unpack into _/server/plugins/myplugin_1.0.0.0_

This checks to see if the original path is valid, and if not, then appends the version number at the end.

Reply:
This snippit is from the configuration file path, it looks like what you're speaking of is in the constructor

Reply:
Couldn't it just be typeof(this).Assembly.Location if we want it to return the path of the plugin?

Reply:
Simplified and corrected.

Reply:
ignore me i was reading PluginConfigurationsPath as PluginPath.

Reverted change.